### PR TITLE
Splines kokkos kernels backend

### PR DESCRIPTION
**Pull request by blegouix**

### CODE REVIEW

File: .github/workflows/tests.yml
Comment (blegouix) [inline .github/workflows/tests.yml:150]: I temporarely remove the `-Werror=extra-semi old-style-cast"` to fix the CI while we discuss the best solution to have a KokkosKernel installation with @yasahi-hpc functions available.

Comment (yasahi-hpc) [reply .github/workflows/tests.yml:150]: @tpadioleau  wants us to update CI to pre-install KK to avoid this error. For me, it is OK as a temporary solution

Comment (blegouix) [reply .github/workflows/tests.yml:150]: Ok I do it or you do ?

Comment (yasahi-hpc) [reply .github/workflows/tests.yml:150]: If you can do that, it would be great. I would like to work on KK side for further fixes, more functionalities (teams), etc..

Comment (blegouix) [reply .github/workflows/tests.yml:150]: Okay if you can just take a look at the CI, I think the error comes from a missing [[maybe_unused]] in your code.

Comment (yasahi-hpc) [reply .github/workflows/tests.yml:150]: Sure. I will fix that after tbsv is merged to KK. Or you would like a fix as soon as possible?

Comment (blegouix) [reply .github/workflows/tests.yml:150]: No it's not urgent at all

File: .github/workflows/tests.yml
Comment (tpadioleau) [inline .github/workflows/tests.yml:204]:
```suggestion
            -DKokkosKernels_ENABLE_COMPONENT_BLAS=OFF \
```

File: .github/workflows/tests.yml
Comment (blegouix) [reply .github/workflows/tests.yml]: @yasahi-hpc can you take a look at the CI ? Thomas made a commit to install KK but there is a problem related to:
https://github.com/yasahi-hpc/kokkos-kernels/blob/6a9535d8200054484fc4158bebb5d8673a6318cf/batched/dense/impl/KokkosBatched_Gbtrs_Serial_Impl.hpp#L21

Maybe remove the `../src/` ?

Comment (yasahi-hpc) [reply .github/workflows/tests.yml]: True. 
I will make a fix soon

**Comment by tpadioleau** [comment]
@yasahi-hpc If you have time can you have a look at the following error
```bash
[  6%] Building CXX object examples/CMakeFiles/characteristics_advection.dir/characteristics_advection.cpp.o
In file included from /data/opt/kokkos-kernels/include/KokkosBatched_Gbtrs.hpp:62,
                 from /src/include/ddc/kernels/splines/splines_linear_problem_band.hpp:19,
                 from /src/include/ddc/kernels/splines/splines_linear_problem_maker.hpp:12,
                 from /src/include/ddc/kernels/splines/spline_builder.hpp:15,
                 from /src/include/ddc/kernels/splines.hpp:17,
                 from /src/examples/characteristics_advection.cpp:12:
/data/opt/kokkos-kernels/include/KokkosBatched_Gbtrs_Serial_Impl.hpp:21:10: fatal error: ../src/KokkosBatched_Tbsv.hpp: No such file or directory
   21 | #include "../src/KokkosBatched_Tbsv.hpp"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
```

My guess is that the path is not valid when the library gets installed ?

**Comment by yasahi-hpc** [comment]
@tpadioleau 
I made a fix. Could you please test again

**Comment by tpadioleau** [comment]
> @tpadioleau I made a fix. Could you please test again

What branch are we based on ?

**Comment by blegouix** [comment]
> > @tpadioleau I made a fix. Could you please test again
> 
> What branch are we based on ?
> 
> I guess @blegouix just did the update but I am still interested in the answer



> > @tpadioleau I made a fix. Could you please test again
> 
> What branch are we based on ?
> 
> I guess @blegouix just did the update but I am still interested in the answer

`develop-spline-kernels` from the Yuuichi fork.

**Comment by yasahi-hpc** [comment]
Failures. I guess I need to include `KokkosBatched_Tbsv.hpp` in `KokkosBatched_Gbtrs_Serial_Impl.hpp`

**Comment by tpadioleau** [comment]
> Failures. I guess I need to include `KokkosBatched_Tbsv.hpp` in `KokkosBatched_Gbtrs_Serial_Impl.hpp`

I let you see about it, if you get a circular dependency you can also use forward declarations.

**Comment by yasahi-hpc** [comment]
It should be OK now.

> I let you see about it, if you get a circular dependency you can also use forward declarations.

We do not have a circular dependency here. I just included a wrong file

**Comment by tpadioleau** [comment]
all green!

File: CMakeLists.txt
Comment (blegouix) [inline CMakeLists.txt:264]:
```suggestion
    set(KokkosKernels_ENABLE_TPL_BLAS OFF)
```

File: .github/workflows/tests.yml
Comment (blegouix) [reply .github/workflows/tests.yml:207]: I works I had it on for testing other parts of KK and forgot to turn it off.

**Comment by blegouix** [comment]
@tpadioleau should I rename `ddc::SplineSolver::LAPACK` to `ddc::SplineSolver::KokkosKernels` ?

**Comment by blegouix** [comment]
@yasahi-hpc I revert your commit after our discussion, we do not need push/pop with nvtx-connector

File: vendor/kokkos-kernels
Comment (tpadioleau) [inline vendor/kokkos-kernels:1]: This needs to be discussed with the Gysela team before merging. They rely on a different version of KokkosKernels, this might create a conflict.

I can think of two solutions:
1. Koliop also rebases on Yuuichi's fork
2. we copy the sources of Kokkos Kernels from Yuuichi's fork and only rely on the version 4.3.01

File: include/ddc/kernels/splines/splines_linear_problem_pds_tridiag.hpp
Comment (tpadioleau) [inline include/ddc/kernels/splines/splines_linear_problem_pds_tridiag.hpp:134]: Let's replace `KOKKOS_CLASS_LAMBDA` by a standard `KOKKOS_LAMBDA`.

File: include/ddc/kernels/splines/splines_linear_problem_pds_band.hpp
Comment (tpadioleau) [inline include/ddc/kernels/splines/splines_linear_problem_pds_band.hpp:141]: Let's replace `KOKKOS_CLASS_LAMBDA` by a standard `KOKKOS_LAMBDA`.

File: include/ddc/kernels/splines/splines_linear_problem_band.hpp
Comment (tpadioleau) [inline include/ddc/kernels/splines/splines_linear_problem_band.hpp:194]: Let's replace `KOKKOS_CLASS_LAMBDA` by a standard `KOKKOS_LAMBDA`.

File: include/ddc/kernels/splines/splines_linear_problem_band.hpp
Comment (tpadioleau) [inline include/ddc/kernels/splines/splines_linear_problem_band.hpp:183]: Let's replace `KOKKOS_CLASS_LAMBDA` by a standard `KOKKOS_LAMBDA`.

File: include/ddc/kernels/splines/splines_linear_problem_pds_tridiag.hpp
Comment (tpadioleau) [inline include/ddc/kernels/splines/splines_linear_problem_pds_tridiag.hpp:129]:
```suggestion
        auto e = Kokkos::subview(q_device, 1, Kokkos::pair<int, int>(0, q_device.extent_int(1) - 1));
```

**Review by tpadioleau — CHANGES_REQUESTED** [review]
Seems globally fine to me!

File: vendor/kokkos-kernels
Comment (yasahi-hpc) [reply vendor/kokkos-kernels:1]: If you have any specific preference in KK version, I can rebase my fork on top of it

Comment (tpadioleau) [reply vendor/kokkos-kernels:1]: > In https://gitlab.maisondelasimulation.fr/gysela-developpers/gyselalibxx/-/merge_requests/550/diffs#00db9f3e96943514c79a8a9a6af0b97240ef0be0
> 
> I use the Yuuichi's fork in gysela and have no manifest problem with Koliop.

Yes but this implies that we cannot use a standard KokkosKernels already installed for both Gysela and DDC. All people should agree on this then.

Comment (blegouix) [reply vendor/kokkos-kernels:1]: Yes I agree, this was just a precision

Comment (yasahi-hpc) [reply vendor/kokkos-kernels:1]: Agree. We should discuss this tomorrow

File: CMakeLists.txt
Comment (tpadioleau) [inline CMakeLists.txt:271]: One last thing, we will need to handle this new dependency as Kokkos. We need to introduce an option `DDC_KokkosKernels_DEPENDENCY_POLICY`.

Comment (blegouix) [reply CMakeLists.txt]: addressed in [7ff91e0](https://github.com/CExA-project/ddc/pull/480/commits/7ff91e0aab724e982fdc0f8a06c6900c8799a553)

File: CMakeLists.txt
Comment (tpadioleau) [inline CMakeLists.txt:282]:
```suggestion
    find_package(KokkosKernels REQUIRED)
```

**Comment by tpadioleau** [comment]
Approved, stalled because of coordination with the Gysela team.

File: CMakeLists.txt
Comment (yasahi-hpc) [inline CMakeLists.txt:256]: Is it possible to switch ginkgo and KK backend with CMake options?

Comment (blegouix) [reply CMakeLists.txt:256]: No, atm we need both to compile the splines. Maybe for another MR

Comment (yasahi-hpc) [reply CMakeLists.txt:256]: Sure

**Comment by blegouix** [comment]
@tpadioleau need the real approval, maybe take a last look at [beb1ee9](https://github.com/CExA-project/ddc/pull/480/commits/beb1ee90e0ada1171cd299fd1705441936caf9ff) and the splines order 1 issue

**Review by tpadioleau — APPROVED** [review]
Ok for me. Maybe wait for Gysela to have updated to Ginkgo 1.8.0 before merging ?

**Comment by yasahi-hpc** [comment]
Let me confirm that GYSELA has a control to choose ginkgo or KK as a spline solver. 
If this is the case, I think this PR is ready

**Comment by blegouix** [comment]
> Ok for me. Maybe wait for Gysela to have updated to Ginkgo 1.8.0 before merging ?

As you want but I see no reason for that. It avoids another ddc sync which has no impact on gysela.

> Let me confirm that GYSELA has a control to choose ginkgo or KK as a spline solver. If this is the case, I think this PR is ready

Yes, LAPACK or GINKGO is specified in template argument of `SplineBuilder`s

**Pull request merged**
